bin/tx-prover/src/utils.rs

```diff
@@ -1,18 +1,109 @@
 use std::time::Duration;
 
+use opentelemetry::{trace::TracerProvider as _, KeyValue};
+use opentelemetry_sdk::{
+    runtime,
+    trace::{RandomIdGenerator, Sampler, TracerProvider},
+    Resource,
+};
+use opentelemetry_semantic_conventions::{
+    resource::{SERVICE_NAME, SERVICE_VERSION},
+    SCHEMA_URL,
+};
 use pingora::{http::ResponseHeader, Error, ErrorType};
 use pingora_proxy::Session;
 use tonic::transport::Channel;
 use tonic_health::pb::health_client::HealthClient;
+use tracing_subscriber::{layer::SubscriberExt, Registry};
+
+pub const MIDEN_TX_PROVER: &str = "miden-tx-prover";
 
 const RESOURCE_EXHAUSTED_CODE: u16 = 8;
 
-pub(crate) fn setup_tracing() {
-    // Set a default log level if `RUST_LOG` is not set
-    if std::env::var("RUST_LOG").is_err() {
-        std::env::set_var("RUST_LOG", "info"); // Default to 'info' level
-    }
-    tracing_subscriber::fmt::init();
+/// Configures and initializes an OpenTelemetry [TracerProvider] with an OTLP exporter.
+///
+/// This function sets up a [TracerProvider] to collect, process, and export spans to an
+/// OTLP-compatible backend (e.g., OpenTelemetry Collector) using the following configuration:
+///
+/// ### Configuration Details
+/// - **OTLP Exporter**:
+///   - The [opentelemetry_otlp::SpanExporter] is configured with the gRPC-based OTLP protocol via
+///     the `tonic` library.
+///   - It ensures spans are exported to an OTLP endpoint.
+///   - Psanics if it fails to initialize.
+/// - **Sampler**:
+///   - A [Sampler::ParentBased] sampler is used, inheriting the parent's sampling decision for
+///     child spans.
+///   - The root spans use a [Sampler::TraceIdRatioBased] sampler with a 100% sampling rate (`1.0`).
+/// - **ID Generator**:
+///   - A `RandomIdGenerator` is used to generate unique trace and span IDs.
+/// - **Resource**:
+///   - A custom resource is created using the `create_resource` function, including attributes such
+///     as the service name, version and schema URL.
+/// - **Batch Exporter**:
+///   - The spans are exported in batches for improved performance, using the Tokio runtime.
+///
+/// ### Returns
+/// - `Ok(TracerProvider)`: If the tracer provider is successfully initialized.
+/// - `Err(String)`: If the OTLP exporter fails to initialize, an error message is returned.
+pub(crate) fn init_tracer_provider() -> Result<TracerProvider, String> {
```

Comment:
Does this function need to be `pub(crate)`? It seems like we use it only internally in this module.

If the above is correct, it is still not clear to me why we don't just combine it with `setup_tracing()` function (i.e., move the code here into `setup_tracing()` function).